### ISSUE TITLE
Alignments from lattices: why lattice-best-path, not lattice-oracle?

### ISSUE THREAD

**Issue by kkm000** [kaldi-asr/kaldi#3242]
The alignment scripts (steps/align_fmllr_lats.sh, steps/nnet3/align_lats.sh) are using ` lattice-best-path` to make alignments from the newly produced lattices. I remember even adjusting the `--acoustic-scale` to match the best test decode, to get a bit cleaner alignments (they were not amazingly important, but since I had the number, why not). It kinda just struck me just now, should they rather use `lattice-oracle`, so that the ground truth transcriptions are used?

I did not try yet, mostly just asking, maybe there is a reason this is not the right idea that I am not aware of.
https://github.com/kaldi-asr/kaldi/blob/f8021d7650bec72c5b3fa8f5d0155bd83f79f04b/egs/wsj/s5/steps/align_fmllr_lats.sh#L158-L163

(and, come think of it, the core strep of these scripts, `*-latgen-faster`, is not using the ground truth either).

----

Bottom line: add a check for acoustic_weight==1.0 for a chain system in steps/nnet3/align_lats.sh.

**Comment by danpovey** [comment]
The meaning of alignment is that we are aligning the transcript.  So the graphs are created from the transcripts; all paths through those lattices would have zero errors.

**Comment by kkm000** [comment]
I know they should, given the wide lattice and decoding beam. But it had been nagging me in the back of my mind that I saw a lot of messages in logs when the decode was restarting and halving the beam when its state space  was exceeding the memory limit.

> WARNING (nnet3-latgen-faster[5.5.210-bf33f]:CheckMemoryUsage():determinize-lattice-pruned.cc:316) Did not reach requested beam in determinize-lattice: size exceeds maximum 50000000 bytes; (repo,arcs,elems) = (34355808,136288,15520896), after rebuilding, repo size was 31476704, effective beam was 4.08656 vs. requested beam 20  
>
> WARNING (nnet3-latgen-faster[5.5.210-bf33f]:DeterminizeLatticePruned():determinize-lattice-pruned.cc:1281) Effective beam 4.40847 was less than beam 20 * cutoff 0.5, pruning raw lattice with new beam 10 and retrying.

"A lot" as in on the order of number of utterances. It also takes an enormous run time. The shortest path in this lattice may not necessary be the theoretically best any more? And the acoustic scale is a free parameter here that I cannot really make sense of--there should be _the_ alignment of a transcript given the model, or am I really deeply confused? And it seems that the decode should not blow up so wide in the first place (and this is a typical chain model, i. e. bigram and nearly HMM-free, output dim about 2100, quite narrow). The result is acceptable, the lattices and the alignments work fine down the line, but something just does not feels right.

**Comment by danpovey** [comment]
You have to set --acoustic-scale 1.0 for a chain system.  I notice the
script and checks for --scale-opts when it's a chain system but not
acoustic-scale.

**Comment by kkm000** [comment]
Looks like I did not, thank you Dan, spot on!

**Comment by danpovey** [comment]
Thanks, please modify the script to also warn about that.

**Comment by kkm000** [comment]
Sure. I'll reopen this to track.

**Comment by kkm000** [comment]
Ah, now we're churnin'! 7.5 minutes vs 2 hours, no warnings.

**Comment by kkm000** [comment]
@danpovey, I just want to share a data point I hope you may find interesting. I compared decodes from a 4 way unconventionally perturbed × 100 hour model that I develop with, using 2 of our test sets, which come here by OK (self-explanatory; the 2nd best), and XD (extremely dirty, acoustically; the worst). I am currently experimenting with this mid-size model so I can get results overnight. The model is the latest "standard" chain-objective skip-layer TDNN-F type, and I'm using constrained=false, so it probably has more alignment freedom.

The exact same training run with the botched alignments gave me a model with the WER of 8.7% on the OK set and 45% on the XD. With the correct alignments, after you pointed out the error, the WER was 8.0% on the OK and 40% on the XD sets. The holy grail of the XD set is maybe 35%; I've never seen better that 36%. 7-9% on the OK set is in the ballpark for a not super fine tuned 100-hr training.

So the unexpected (at least to me) effect is that the excessively deep alignment lattices used for training produced a model with a statistically significant but quite small degradation in decoding cleaner data, but knocked its performance on a dirty input set very far down. Both models had a very similar convergence dynamic during training, and the end objectives were not dramatically different, either:
```
borked ali: xent:train/valid[133,201]=(-0.934,-0.715/-1.01,-0.885) logprob:train/valid[133,201]=(-0.053,-0.029/-0.064,-0.060)
fixed ali:  xent:train/valid[133,201]=(-0.920,-0.710/-0.978,-0.875) logprob:train/valid[133,201]=(-0.053,-0.030/-0.064,-0.060)
```
I do not have any intuition for this, and am not asking a question. I just wanted to share this tidbit with you, hoping you might find it useful: you know so much about this stuff that this piece may click in with your mental picture. Heck, in less flowery words: I won't get a zilch of useful new knowledge from this my blooper, you likely will. :)

**Comment by danpovey** [comment]
It could be there is a ceiling on performance around 4% due to genuine
ambiguity and transcriber error, and those two systems actually have
similar relative degradation when measured against that modified floor.
